File: RayTracerModel/Vector3.cs
using System;

namespace RayTracer
{
    public class Vector3
    {
        public double X { get; private set; }
        public double Y { get; private set; }
        public double Z { get; private set; }

        public Vector3(double x, double y, double z)
        {
            X = x;
            Y = y;
            Z = z;
        }

        public double Length()
        {
            return Math.Sqrt(X * X + Y * Y + Z * Z);
        }

        public Vector3 Normalized()
        {
            return new Vector3(X / Length(), Y / Length(), Z / Length());
        }

        public Vector3 Plus(Vector3 other)
        {
            return new Vector3(X + other.X, Y + other.Y, Z + other.Z);
        }
        public Vector3 Minus(Vector3 other)
        {
            return new Vector3(X - other.X, Y - other.Y, Z - other.Z);
        }

        public Vector3 ScalarMultiply(double scalar)
        {
            return new Vector3(X * scalar, Y * scalar, Z * scalar);
        }

        public double DotProduct(Vector3 other)
        {
            return X * other.X + Y * other.Y + Z * other.Z;
        }

        public Vector3 CrossProduct(Vector3 other)
        {
            return new Vector3(
                Y * other.Z - Z * other.Y,
                Z * other.X - X * other.Z,
                X * other.Y - Y * other.X);
        }

        public override bool Equals(object obj)
        {
            if (obj.GetType() != typeof(Vector3)) return false;
            var v = obj as Vector3;
            return (v.X == X) && (v.Y == Y) && (v.Z == Z);
        }

        public override int GetHashCode()
        {
            return X.GetHashCode() + Y.GetHashCode() + Z.GetHashCode();
        }

        public static Vector3 operator +(Vector3 a, Vector3 b)
        {
            return a.Plus(b);
        }

        public static Vector3 operator -(Vector3 a, Vector3 b)
        {
            return a.Minus(b);
        }

        public static Vector3 operator *(Vector3 a, Vector3 b)
        {
            return a.CrossProduct(b);
        }

        public static Vector3 operator *(Vector3 a, double b)
        {
            return a.ScalarMultiply(b);
        }
        public static Vector3 operator *(double a, Vector3 b)
        {
            return b.ScalarMultiply(a);
        }
    }
}

File: RayTracerVBSingleProj/Vector3.vb
Public Class Vector3
    Public Property X As Double
    Public Property Y As Double
    Public Property Z As Double

    Public Sub New(ByVal x As Double, ByVal y As Double, ByVal z As Double)
        Me.X = x
        Me.Y = y
        Me.Z = z
    End Sub

    Public Function Length() As Double
        Return Math.Sqrt(X * X + Y * Y + Z * Z)
    End Function

    Public Function Normalized() As Vector3
        Return New Vector3(X / Length(), Y / Length(), Z / Length())
    End Function

    Public Function Plus(ByVal other As Vector3) As Vector3
        Return New Vector3(X + other.X, Y + other.Y, Z + other.Z)
    End Function

    Public Function Minus(ByVal other As Vector3) As Vector3
        Return New Vector3(X - other.X, Y - other.Y, Z - other.Z)
    End Function

    Public Function ScalarMultiply(ByVal scalar As Double) As Vector3
        Return New Vector3(X * scalar, Y * scalar, Z * scalar)
    End Function

    Public Function DotProduct(ByVal other As Vector3) As Double
        Return X * other.X + Y * other.Y + Z * other.Z
    End Function

    Public Function CrossProduct(ByVal other As Vector3) As Vector3
        Return New Vector3(Y * other.Z - Z * other.Y, Z * other.X - X * other.Z, X * other.Y - Y * other.X)
    End Function

    Public Overrides Function Equals(ByVal obj As Object) As Boolean
        If obj.GetType() <> GetType(Vector3) Then Return False
        Dim v = TryCast(obj, Vector3)
        Return (v.X = X) AndAlso (v.Y = Y) AndAlso (v.Z = Z)
    End Function

    Public Overrides Function GetHashCode() As Integer
        Return X.GetHashCode() + Y.GetHashCode() + Z.GetHashCode()
    End Function

    Public Shared Operator +(a As Vector3, b As Vector3) As Vector3
        Return a.Plus(b)
    End Operator

    Public Shared Operator -(a As Vector3, b As Vector3) As Vector3
        Return a.Minus(b)
    End Operator
    Public Shared Operator *(a As Vector3, b As Vector3) As Vector3
        Return a.CrossProduct(b)
    End Operator
    Public Shared Operator *(a As Vector3, b As Double) As Vector3
        Return a.ScalarMultiply(b)
    End Operator
    Public Shared Operator *(a As Double, b As Vector3) As Vector3
        Return b.ScalarMultiply(a)
    End Operator
End Class
